### PR TITLE
Clone collection before foreach in DuplicateKey filter

### LINKED ISSUE #44: DuplicateKey filter removes "every second entry" instead of "all but last"
DuplicateKey filter removes "every second entry" instead of "all but last"

To reproduce:
Make 4 timeline entries with the same DuplicateKey and the same priority. Run the filter and it will leave entries 2 and 4.

IMHO bug is caused because of
https://github.com/stephpy/timeline/blob/master/src/Filter/DuplicateKey.php#L42

```
unset($collection[$keyToDelete]);
```

Here it unsets an item of the foreach collection. PHP does not clone it automatically (as it is probably assumed) and removing the entry skips the next iteration.

I would propose replacing https://github.com/stephpy/timeline/blob/master/src/Filter/DuplicateKey.php#L21

with

```
        foreach ($collection as $key => $value) {
            $cloned[$key] = $value;
        }

        foreach ($cloned as $key => $action) {

### PR DESCRIPTION
**Pull request by venimus**
fixes #44

### CODE REVIEW

**Comment by stephpy** [comment]
Thanks

**Comment by venimus** [comment]
Could you please create a release if you are satisfied

**Comment by stephpy** [comment]
https://github.com/stephpy/timeline/releases/tag/v1.0.4 Done ;)

**Comment by venimus** [comment]
:+1:

File: src/Filter/DuplicateKey.php [hunk 1]
```diff
@@ -17,8 +17,13 @@ class DuplicateKey extends AbstractFilter implements FilterInterface
     public function filter($collection)
     {
         $duplicateKeys = array();
+        $clonedCollection = array();
 
-        foreach ($collection as $key => $action) {
+        foreach ($collection as $key => $value) {
```
Comment (Zeichen32) [inline src/Filter/DuplicateKey.php]: Why do you not use?
```php
$clonedCollection = clone $collection
```

Comment (venimus) [inline src/Filter/DuplicateKey.php]: Sometime it is array

**Comment by Zeichen32** [comment]
Oh okay, then you can use:

I don't know if it is a good or bad practice.
But i think, its maybe a bit faster for large collections :sweat_smile: 

```php
$clonedCollection = is_object($collection) ? clone $collection : $collection;
```

Tests:
```php
$collection = array(1, 2, 3);
$clonedCollection = is_object($collection) ? clone $collection : $collection;

unset($clonedCollection[2]);

var_dump($collection === $clonedCollection);
var_dump($collection);
var_dump($clonedCollection);

// bool(false) 
// array(3) { [0]=> int(1) [1]=> int(2) [2]=> int(3) }
// array(2) { [0]=> int(1) [1]=> int(2) } 
```

```php
$collection = (object) array('foo' => 1, 'bar' => 2);
$clonedCollection = is_object($collection) ? clone $collection : $collection;

unset($clonedCollection->foo);

var_dump($collection === $clonedCollection);
var_dump($collection);
var_dump($clonedCollection);

// bool(false)
// object(stdClass)#1 (2) { ["foo"]=> int(1) ["bar"]=> int(2) } 
// object(stdClass)#2 (1) { ["bar"]=> int(2) }
```

```php
$collection = (object) array(
	'foo' => (object) array('baz' => 1),
	'bar' => (object) array('baz' => 2)
);

$clonedCollection = is_object($collection) ? clone $collection : $collection;

unset($clonedCollection->foo);

var_dump($collection === $clonedCollection);
var_dump($collection);
var_dump($clonedCollection);

// bool(false)
// object(stdClass)#3 (2) { ["foo"]=> object(stdClass)#1 (1) { ["baz"]=> int(1) } ["bar"]=> object(stdClass)#2 (1) { ["baz"]=> int(2) } }
// object(stdClass)#4 (1) { ["bar"]=> object(stdClass)#2 (1) { ["baz"]=> int(2) } }
```

**Comment by Zeichen32** [comment]
Then you can use:
```php
$clonedCollection = is_object($collection) ? clone $collection : $collection;
```
I don't know if it is a good or bad practice, but i think its a bit faster for large collections.

Some tests:

```php
$collection = array(
	'foo' => 1,
	'bar' => 2
);

$clonedCollection = is_object($collection) ? clone $collection : $collection;

unset($clonedCollection['foo']);

var_dump($collection === $clonedCollection);
var_dump($collection);
var_dump($clonedCollection);

// bool(false)
// array(2) { ["foo"]=> int(1) ["bar"]=> int(2) }
// array(1) { ["bar"]=> int(2) }
````

```php
$collection = (object) array(
	'foo' => 1,
	'bar' => 2
);

$clonedCollection = is_object($collection) ? clone $collection : $collection;

unset($clonedCollection->foo);

var_dump($collection === $clonedCollection);
var_dump($collection);
var_dump($clonedCollection);

// bool(false)
// object(stdClass)#1 (2) { ["foo"]=> int(1) ["bar"]=> int(2) }
// object(stdClass)#2 (1) { ["bar"]=> int(2) }
```

```php
$collection = (object) array(
	'foo' => (object) array('baz' => 1),
	'bar' => (object) array('baz' => 2)
);

$clonedCollection = is_object($collection) ? clone $collection : $collection;

unset($clonedCollection->foo);

var_dump($collection === $clonedCollection);
var_dump($collection);
var_dump($clonedCollection);

// bool(false)
// object(stdClass)#3 (2) { ["foo"]=> object(stdClass)#1 (1) { ["baz"]=> int(1) } ["bar"]=> object(stdClass)#2 (1) { ["baz"]=> int(2) } }
// object(stdClass)#4 (1) { ["bar"]=> object(stdClass)#2 (1) { ["baz"]=> int(2) } }
```

**Comment by stephpy** [comment]
:+1:

**Comment by venimus** [comment]
But if it is an array, this solution don't "clone" it, which leads to the original issue.

**Comment by Zeichen32** [comment]
See my test no #1. The Array will be clone, because it's a scalar type.

**Comment by venimus** [comment]
When I tracked the issue my collection was array of objects. Since it contains references, foreach will not clone it automatically as if it was a simple array. I'm sure there are more beautiful implementations, but the speed improvement is questionable because this filter is after pagination. Meaning it will contain mostly small number of items. Furthermore foreach internally is quite optimized, while clone is something very similar as implementation. That's why I preferred a simple to read code. I actually think the whole function needs refactoring anyway, but this is for the project owner to decide.

**Pull request merged**